byron/ledger/executable-spec/test/Ledger/HasTypeReps/Properties.hs

```diff
@@ -0,0 +1,170 @@
+import qualified Data.Map.Strict                    as Map
+import           Data.Sequence                      (empty, (<|), (><))
+import qualified Data.Sequence                      as Seq
+import           Data.Typeable                      (TypeRep, typeOf)
+import           Numeric.Natural                    (Natural)
+
+import           Hedgehog                           (MonadTest, Property,
+                                                     forAll, property,
+                                                     withTests, (===))
+import           Test.Tasty.Hedgehog
+
+import           Control.State.Transition.Generator (trace)
+import           Control.State.Transition.Trace     (TraceOrder (OldestFirst),
+                                                     traceSignals)
+
+import           Ledger.Core                        hiding ((<|))
+import           Ledger.UTxO
+
+import           Cardano.Ledger.Spec.STS.UTXOW      (UTXOW)
+
+import           Test.Tasty                         (TestTree, testGroup)
+import           Test.Tasty.HUnit                   (Assertion, testCase, (@?=))
+
+--------------------------------------------------------------------------------
+-- Example HasTypeReps.typeReps for TxIn, Tx
+--------------------------------------------------------------------------------
+
+aTx :: Tx
+aTx = undefined
+
+aTxId :: TxId
+aTxId = TxId (hash aTx)
+
+-- | 'TxIn' has a generic instance for 'HasTypeReps', which recursively adds
+--   'typeReps' for all types within 'TxIn'.
+exampleTypeRepsTxIn :: Assertion
+exampleTypeRepsTxIn =
+  let txIn = TxIn aTxId 0
+  in (typeReps txIn) @?= typeOf (undefined::TxIn)
+                         <| typeOf (undefined::TxId)
+                         <| typeOf (undefined::Hash)
+                         <| typeOf (undefined::Int)
+                         <| typeOf (undefined::Natural)
+                         <| empty
+
+-- | A 'TxWits' term may contain multiple inputs/outputs/witnesses.
+--   In this example, we have 2 inputs and show how the 'typeReps' for
+--   'TxIn' is repeated twice.
+exampleTypeRepsTx :: Assertion
+exampleTypeRepsTx =
+  let
+    (in0,in1) = (TxIn aTxId 0, TxIn aTxId 1)
+    outs = []
+    wits = []
+    tx = TxWits (Tx [in0, in1] outs) wits
+  in (typeReps tx) @?= typeOf (undefined::TxWits)
+                       <| typeOf (undefined::Tx)
+                       <| typeOf (undefined::[TxIn])
+                       <| typeReps in0
+                       >< typeReps in1
+                       >< (Seq.fromList [ typeOf (undefined::[TxOut])
+                                        , typeOf (undefined::[Wit]) ])
+
+--------------------------------------------------------------------------------
+-- Properties of abstractSize of TxWits / TxIn /TxOut / Wit
+--------------------------------------------------------------------------------
+
+-- | Since 'VKey' appears in both Wit and TxOut (as part of Addr), we
```

Comment:
Would it make sense to clarify why did we choose different costs?

Reply:
👍